pkg/vsphere/disk/disk_manager.go

```diff
@@ -275,24 +281,38 @@ func (m *Manager) Detach(op trace.Operation, d *VirtualDisk) error {
 		return errors.Trace(err)
 	}
 
-	spec := types.VirtualMachineConfigSpec{}
-
-	disk, err := findDisk(op, m.vm, d.DatastoreURI)
+	disk, err := findDiskByFilename(op, m.vm, d.DatastoreURI)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
+	if err = m.detach(op, disk); err != nil {
+		op.Errorf("detach for %s failed with %s", d.DevicePath, errors.ErrorStack(err))
+		return errors.Trace(err)
+	}
+
+	select {
```

Comment:
I wander why do you even need this select. are you trying to drain the channel?
if len(m.maxAttached) > 0 {
 <- m.maxAttached
}

It is MUCH MORE lightweight comparable to a select. Go Select implementation is around 600 lines. The only thing that does't work in my example is that it is not concurrent safe. you potentially may block forever if two calls happen simultaneously. I leave it up to you to decide.